library/src/main/kotlin/com/vinted/actioncable/client/kotlin/Connection.kt

```diff
@@ -38,7 +39,8 @@ class Connection constructor(
             var reconnectionMaxAttempts: Int = 30,
             var reconnectionDelay: Int = 3,
             var reconnectionDelayMax: Int = 30,
-            var okHttpClientFactory: OkHttpClientFactory? = null
+            var okHttpClientFactory: OkHttpClientFactory? = null,
+            var okHttpWebSocketFactory: WebSocket.Factory? = null
```

Comment:
Hi, thanks for the PR!

Would it be possible to have single option field for providing `WebSocket.Factory`? As current solution would allow to provide both `okHttpClientFactory` and `okHttpWebSocketFactory` and silently `okHttpWebSocketFactory` is preferred. I think this would be confusing library API.

Reply:
Hi @GediminasZukas, thanks for the review!

Just so I understand correctly, you would prefer if the client could configure just one field for `OkHttpClientFactory` and `WebSocket.Factory`? Please correct me if I've misunderstood.

It's not possible to provide a custom `WebSocket.Factory` by configuring `OkHttpClient`. You also cannot extend `OkHttpClient`, since actioncable-client-kotlin will call `newBuilder()`:

https://github.com/vinted/actioncable-client-kotlin/blob/c8765676818d9e4d73e3ea26f6664df6e464fccf/library/src/main/kotlin/com/vinted/actioncable/client/kotlin/Connection.kt#L124-L125

This overrides any customization the client made (calling `build()` will return `new OkHttpClient()`. Also, clients are discouraged from extending `OkHttpClient`, I checked with OkHttp here: https://github.com/square/okhttp/issues/6242

I believe the only solution is to provide a separate field. I had the same PR with Ktor, here: https://github.com/ktorio/ktor/pull/2035

Please let me know if you have some suggestion. Thanks.

Reply:
> Just so I understand correctly, you would prefer if the client could configure just one field for OkHttpClientFactory and WebSocket.Factory? Please correct me if I've misunderstood.

Yes, you are basically correct. Actually I am thinking that maybe would be possible to move full building of `WebSocket.Factory` implementation to the client side instead of doing it partially internally. Here is the idea how it could look like:

**Library:**

single option field for factory: `var websocketFactory: WebSocket.Factory`

```
val websocketFactory = (options.websocketFactory?.invoke() 
         ?: OkHttpClient()).newBuilder().sslSocketFactory(it.socketFactory).hostnameVerifier(it).build() // no custom factory provided, build client internally with supplied option params
```

**Client:**

```
options.connection.websocketFactory = {
    OkHttpClient().newBuilder()
                    .addNetworkInterceptor(StethoInterceptor()) 
                    .sslSocketFactory(it.socketFactory) // mention in docs to set directly on custom client
                    .hostnameVerifier(it) // mention in docs to set directly on custom client
                    .build()
}
```

or

```
options.connection.websocketFactory = {
    object : WebSocket.Factory {
        // your custom implementation
     }
}
```

Would such implementation solve the problem?

Reply:
If I understand correctly, what you propose would be to change the connection options, so that the current field:

```
// OkHttpClientFactory is typealias for () -> OkHttpClient
var okHttpClientFactory: OkHttpClientFactory?
```
would become a new field:
```
var webSocketFactory: () -> WebSocket.Factory
```
And we would remove the old `okHttpClientFactory` field? And we would update docs to tell clients they need to configure OkHttpClient in their code.

This is fine for me, but would be a breaking change for other users. Is that ok with you?

Reply:
Yes, you are correct. I think we can drop that typealias too. Also lambda type seems not needed, probably could be just: `var webSocketFactory: WebSocket.Factory`

> This is fine for me, but would be a breaking change for other users. Is that ok with you?

This library is more less is in an early stage. I think this API change wouldn't be big. Also, we could make small migration guide in release notes.

@neworld , what do you think about all of this?

Reply:
I like this idea as it adds more transparency.

Reply:
Ok I will update with these changes!

Reply:
Updated now... please review closely to be sure I haven't removed anything I shouldn't remove, and made the correct changes. Thanks.